File: code/python/medication.py
from dataclasses import dataclass

@dataclass
class TabletDosage:
    morning: int
    midday: int
    evening: int

@dataclass
class InfusionDosage:
    speed: float
    duration: int

# Requires python 3.12
type Dosage = TabletDosage | InfusionDosage

@dataclass
class Medication:
    drugName: str
    dosage: Dosage

def format(m: Medication) -> str:
    return f'{m.drugName}: {formatDosage(m.dosage)}'

def formatDosage(d: Dosage) -> str:
    match d:
        case TabletDosage():
            return f'{d.morning}-{d.midday}-{d.evening}'
        case InfusionDosage():
            return f'{d.speed} ml/min for {d.duration}h'

paracetamol = Medication("Paracetamol", TabletDosage(1,0,2))
infliximab = Medication("Infliximab", InfusionDosage(1.5, 2))
print(format(paracetamol))
print(format(infliximab))

File: code/csharp/Medication.cs
public record Dosage {
    public record Tablet(int morning, int midday, int evening) : Dosage();
    public record Infusion(double speed, int duration) : Dosage();

    public string format() {
        return this switch {
            Tablet t => t.morning + "-" + t.midday + "-" + t.evening,
            Infusion i => i.speed + "ml/min for " + i.duration + "h",
            _ => throw new ApplicationException("unexpected dosage: " + this)
        };
    }

    private Dosage() {} // private constructor can prevent derived cases from being defined elsewhere
}

public record Medication(string drugName, Dosage dosage) {

    public string format() {
        return this.drugName + ": " + this.dosage.format();
    }
}
